src/wp-includes/webfonts-api/class-wp-webfonts-controller.php

```diff
@@ -0,0 +1,252 @@
+	}
+
+	/**
+	 * Gets the instance of the providers' registry.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return WP_Webfonts_Provider_Registry
+	 */
+	public function providers() {
+		return $this->providers_registry;
+	}
+
+	/**
+	 * Generate and enqueue webfonts styles.
+	 *
+	 * @since 5.9.0
+	 */
+	public function generate_and_enqueue_styles() {
+		// Generate the styles.
+		$styles = $this->generate_styles();
+
+		// Bail out if there are no styles to enqueue.
+		if ( '' === $styles ) {
+			return;
+		}
+
+		// Enqueue the stylesheet.
+		wp_register_style( $this->stylesheet_handle, '' );
+		wp_enqueue_style( $this->stylesheet_handle );
+
+		// Add the styles to the stylesheet.
+		wp_add_inline_style( $this->stylesheet_handle, $styles );
+	}
+
+	/**
+	 * Generate and enqueue editor styles.
+	 *
+	 * @since 5.9.0
+	 */
+	public function generate_and_enqueue_editor_styles() {
+		// Generate the styles.
+		$styles = $this->generate_styles();
+
+		// Bail out if there are no styles to enqueue.
+		if ( '' === $styles ) {
+			return;
+		}
+
+		wp_add_inline_style( 'wp-block-library', $styles );
+	}
+
+	/**
+	 * Generate styles for webfonts.
+	 *
+	 * By default (due to privacy concerns), this API will not do remote requests to
+	 * external webfont services nor generate `@font-face` styles for these remote
+	 * providers. The filter `'has_remote_webfonts_request_permission'` is provided
+	 * to grant permission to do the remote request.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string $styles Generated styles.
+	 */
+	private function generate_styles() {
+		$providers = $this->providers_registry->get_all_registered();
+		if ( empty( $providers ) ) {
+			return '';
+		}
+
+		$styles = '';
+
+		/*
+		 * Loop through each of the providers to get the CSS for their respective webfonts
+		 * to incrementally generate the collective styles for all of them.
+		 */
+		foreach ( $providers as $provider_id => $provider ) {
+			$registered_webfonts = $this->webfonts_registry->get_by_provider( $provider_id );
+
+			// If there are no registered webfonts for this provider, skip it.
+			if ( empty( $registered_webfonts ) ) {
+				continue;
+			}
+
+			/*
+			 * Skip fetching from a remote fonts service if the user has not
+			 * consented to the remote request.
+			 */
+			if (
+				'local' !== $provider_id &&
+				/**
+				 * Allows permission to be set for doing remote requests
+				 * to a webfont service provider.
+				 *
+				 * By default, the Webfonts API will not make remote requests
+				 * due to privacy concerns.
+				 *
+				 * @since 5.9.0
+				 *
+				 * @param bool  $has_permission Permission to do the remote request.
+				 *                              Default false.
+				 * @param string $provider_id   Provider's ID, e.g. 'google', to identify
+				 *                              the remote webfonts service provider.
+				 */
+				true !== apply_filters( 'has_remote_webfonts_request_permission', false, $provider_id )
+			) {
+				continue;
+			}
```

Comment:
Hey @azaozz, here's the new filter we discussed for granting permission to do a remote request to an external webfonts service provider, such as Google Fonts. By default, it's set to `false` meaning the API will not do remote requests. What do you think?

Reply:
Huzzah